<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>API Reference - DEPy</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "API Reference";
        var mkdocs_page_input_path = "api.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> DEPy
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../usage/">Usage</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">API Reference</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#depy.summarized_py">summarized_py</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.filter_features">filter_features</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.filter_missingness">filter_missingness</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.filter_samples">filter_samples</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.import_from_delim_file">import_from_delim_file</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.impute_missing_values">impute_missing_values</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.limma_trend_dea">limma_trend_dea</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.load_example_data">load_example_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.load_sp">load_sp</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.plot_pca">plot_pca</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.save_sp">save_sp</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.select_variable_features">select_variable_features</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.surrogate_variable_analysis">surrogate_variable_analysis</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.transform_features">transform_features</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#depy.summarized_py.SummarizedPy.volcano_plot">volcano_plot</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#depy.column_selector">column_selector</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#depy.column_selector.ColumnSelector">ColumnSelector</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#depy.column_selector.ColumnSelector.select_cols">select_cols</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../HISTORY/">Changelog</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">DEPy</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">API Reference</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/SebastianDHA/DEPy/edit/master/docs/api.md">Edit on SebastianDHA/DEPy</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="api-reference">API Reference<a class="headerlink" href="#api-reference" title="Permanent link">&para;</a></h1>


<div class="doc doc-object doc-module">



<a id="depy.summarized_py"></a>
    <div class="doc doc-contents first">










  <div class="doc doc-children">









<div class="doc doc-object doc-class">



<h2 id="depy.summarized_py.SummarizedPy" class="doc doc-heading">
            <code>SummarizedPy</code>


<a href="#depy.summarized_py.SummarizedPy" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">



        <p>A class to hold bulk proteomics (and metabolomics) data and process it for differential expression analysis.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>data</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A 2D array with shape=(features, samples) holding numerical intensity data.</p>
              </div>
            </li>
            <li>
              <b><code>features</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A DataFrame holding feature metadata.</p>
              </div>
            </li>
            <li>
              <b><code>samples</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A DataFrame holding sample metadata.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Attributes:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>data</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>A 2D array with shape=(features, samples) holding numerical intensity data.</p>
              </div>
            </li>
            <li>
              <b><code>features</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>A DataFrame holding feature metadata.</p>
              </div>
            </li>
            <li>
              <b><code>samples</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>A DataFrame holding sample metadata.</p>
              </div>
            </li>
            <li>
              <b><code>history</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of strings documenting each valid class module call.</p>
              </div>
            </li>
            <li>
              <b><code>results</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>A DataFrame holding DEA results generated by limma_trend_dea method.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If supplied data, features, or samples are incorrect classes.</p>
              </div>
            </li>
            <li>
                  <code><span title="TypeError">TypeError</span></code>
              –
              <div class="doc-md-description">
                <p>If data.shape[0] != features.shape[0] or data.shape[1] != samples.shape[0].</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Constructing SummarizedPy object from numpy array and pandas DataFrame.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>                <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>                <span class="p">[</span><span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">features</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;proteinID&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;feature1&quot;</span><span class="p">,</span> <span class="s2">&quot;feature2&quot;</span><span class="p">,</span> <span class="s2">&quot;feature3&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samples</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;sample&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sample1&quot;</span><span class="p">,</span> <span class="s2">&quot;sample2&quot;</span><span class="p">,</span> <span class="s2">&quot;sample3&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">features</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">)</span>
<span class="go">&lt;SummarizedPy(data=ndarray(shape=(3, 3), dtype=int64), features=DataFrame(shape=(3, 1)), samples=DataFrame(shape=(3, 1)))&gt;</span>
</code></pre></div>










  <div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.filter_features" class="doc doc-heading">
            <code class=" language-python"><span class="n">filter_features</span><span class="p">(</span><span class="n">expr</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.filter_features" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Filter SummarizedPy object based on feature metadata, using either Pandas-like query strings or a mask.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>expr</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A Pandas-style query string that can be interpreted by pd.obj.query(expr=expr).</p>
              </div>
            </li>
            <li>
              <b><code>mask</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A boolean mask for subsetting.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A filtered <code>SummarizedPy</code> object.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If no valid <code>expr</code> or <code>mask</code> argument is supplied.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Filter out reverse hits in example dataset PXD000438.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rev_hits</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="s2">&quot;protein_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">bool</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s2">&quot;REV&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="s2">&quot;rev&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rev_hits</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">filter_features</span><span class="p">(</span><span class="n">expr</span><span class="o">=</span><span class="s2">&quot;~rev&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">filter_features</span><span class="p">(</span><span class="n">mask</span><span class="o">=~</span><span class="n">rev_hits</span><span class="p">)</span> <span class="n">Or</span> <span class="n">using</span> <span class="n">rev_hits</span> <span class="k">as</span> <span class="n">a</span> <span class="n">boolean</span> <span class="n">mask</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.filter_missingness" class="doc doc-heading">
            <code class=" language-python"><span class="n">filter_missingness</span><span class="p">(</span><span class="n">frac</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">strategy</span><span class="o">=</span><span class="s1">&#39;all_conditions&#39;</span><span class="p">,</span> <span class="n">condition_column</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.filter_missingness" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Filter SummarizedPy object based on % feature missingness across one of:
overall, all conditions, or any condition.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>frac</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>0.75</code>
)
              –
              <div class="doc-md-description">
                <p>Minimum percentage valid values. Features with missingness
greater than or equal to (1 - frac) will be excluded.</p>
              </div>
            </li>
            <li>
              <b><code>strategy</code></b>
                  (<code>(<span title="overall">overall</span>, <span title="all_conditions">all_conditions</span>, <span title="any_condition">any_condition</span>)</code>, default:
                      <code>&#39;overall&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Filtering strategy:</p>
<ul>
<li>'overall' : Require &gt;= frac valid values across all samples.</li>
<li>'all_conditions' : Require &gt;= frac valid values in <em>each</em>
  condition defined by <code>condition_column</code>.</li>
<li>'any_condition' : Require &gt;= frac valid values in at least
  one condition defined by <code>condition_column</code>.</li>
</ul>
              </div>
            </li>
            <li>
              <b><code>condition_column</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Name of column in the <code>samples</code> attribute on which to base
filtering, in case of 'all_conditions' or 'any_condition'.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A filtered <code>SummarizedPy</code> object.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If an invalid <code>strategy</code> is supplied or if <code>condition_column</code> is required but missing.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Filter out missing values in example dataset PXD000438.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">filter_missingness</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;overall&quot;</span><span class="p">,</span> <span class="n">frac</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">filter_missingness</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;any_condition&quot;</span><span class="p">,</span> <span class="n">condition_column</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="n">frac</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">filter_missingness</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;all_conditions&quot;</span><span class="p">,</span> <span class="n">condition_column</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="n">frac</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.filter_samples" class="doc doc-heading">
            <code class=" language-python"><span class="n">filter_samples</span><span class="p">(</span><span class="n">expr</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.filter_samples" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Filter SummarizedPy object based on sample metadata, using either Pandas-like query strings or a mask.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>expr</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A Pandas-style query string that can be interpreted by pd.obj.query(expr=expr).</p>
              </div>
            </li>
            <li>
              <b><code>mask</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A boolean mask for subsetting.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A filtered <code>SummarizedPy</code> object.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If no valid <code>expr</code> or mask argument is supplied.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Filter for ADC samples in example dataset PXD000438.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ADC&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">6</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;SCC&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">expr</span><span class="o">=</span><span class="s2">&quot;condition==&#39;ADC&#39;&quot;</span><span class="p">)</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.import_from_delim_file" class="doc doc-heading">
            <code class=" language-python"><span class="n">import_from_delim_file</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">delim</span><span class="p">,</span> <span class="n">data_selector</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">feature_selector</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">replace_val_with_nan</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">clean_column_names</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#depy.summarized_py.SummarizedPy.import_from_delim_file" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Alternative constructor from file.
Import method that reads data directly from delimited file, including feature data,
feature metadata, and sample metadata, and assigns them to data, features, and samples attributes automatically.
This is intended for convenient import of standardized outputs like MaxQuant's proteingroups.txt or
FragPipe/DIA-NN's diann-output.pg_matrix.tsv.
Method uses ColumnSelector objects to assign columns to their relevant storage containers.
If no ColumnSelector is provided, the function defaults to assigning all numerical (float64) columns to data
and all string (object) columns to feature data. Thus, it is best to explicitly state which columns to import.
The samples attribute is automatically populated with the column names from data.
The original data row and columns indices are stored in features and samples 'orig_index' variables, resp., for bookkeeping.
The read path and delimiter used will be appended to the history attribute.
Values in data can be replaced with NaN to indicate missingness (e.g. intensity values of 0).
Column names can be automatically cleaned.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to file to read in.</p>
              </div>
            </li>
            <li>
              <b><code>delim</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Delimiter to parse file (e.g. '     ' for .txt or ',' for .csv).</p>
              </div>
            </li>
            <li>
              <b><code>data_selector</code></b>
                  (<code><a class="autorefs autorefs-internal" title="ColumnSelector (depy.column_selector.ColumnSelector)" href="#depy.column_selector.ColumnSelector">ColumnSelector</a></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A <code>ColumnSelector</code> object with specified names or regex patterns
to extract data columns in file. If not specified,
defaults to all 'number' dtype columns.</p>
              </div>
            </li>
            <li>
              <b><code>feature_selector</code></b>
                  (<code><a class="autorefs autorefs-internal" title="ColumnSelector (depy.column_selector.ColumnSelector)" href="#depy.column_selector.ColumnSelector">ColumnSelector</a></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A <code>ColumnSelector</code> object with specified names or regex patterns
to extract feature metadata columns in file. If not specified,
defaults to all object dtype columns.</p>
              </div>
            </li>
            <li>
              <b><code>replace_val_with_nan</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A float numeric value in data to replace with np.nan to indicate missingness (e.g. 0).</p>
              </div>
            </li>
            <li>
              <b><code>clean_column_names</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to clean column names in file before processing. Note that column selection
happens on the cleaned column names! Thus, you have to account for this
when instantiating the <code>ColumnSelector</code> object. Cleaning will coerce
all string to lower case; spaces and hyphens will be replaced with underscores,
and leading and trailing whitespace will be trimmed.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A <code>SummarizedPy</code> object.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Read in data from protein groups file (e.g. MQ or FragPipe) and construct SummarizedPy object.
Default to placing all numerical columns in 'data', assocaited column names in 'samples', and object or string type columns in 'features'.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pg_path</span> <span class="o">=</span> <span class="s2">&quot;~/path/to/my/proteingroups.txt&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">import_from_delim_file</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="n">pg_path</span><span class="p">,</span> <span class="n">delim</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">replace_val_with_nan</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span> <span class="n">clean_column_names</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
    <p>Select columns to import using ColumnSelector object.
Assume data are in columns containing sub-string 'LFQ_intensity_'.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pg_path</span> <span class="o">=</span> <span class="s2">&quot;~/path/to/my/proteingroups.txt&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">ColumnSelector</span><span class="p">(</span><span class="n">regex</span><span class="o">=</span><span class="s2">&quot;LFQ_intensity_&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">features</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">ColumnSelector</span><span class="p">(</span><span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;proteinID&quot;</span><span class="p">,</span> <span class="s2">&quot;geneSymbol&quot;</span><span class="p">,</span> <span class="s2">&quot;proteinDescription&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">import_from_delim_file</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="n">pg_path</span><span class="p">,</span> <span class="n">delim</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">data_selector</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">feature_selector</span><span class="o">=</span><span class="n">features</span><span class="p">)</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.impute_missing_values" class="doc doc-heading">
            <code class=" language-python"><span class="n">impute_missing_values</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extra_args</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.impute_missing_values" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Impute missing values using the ImputeLCMD R package.</p>
<p>Several common methods are available under the assumptions of:
- MAR (KNN, SVD, MLE)
- MNAR (QRILC, MinDet, MinProb)
- Both MAR and MNAR (Hybrid)</p>
<p>Refer to the ImputeLCMD package documentation for further information:
https://cran.r-project.org/web/packages/imputeLCMD/imputeLCMD.pdf</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>method</code></b>
                  (<code>(<span title="Hybrid">Hybrid</span>, <span title="KNN">KNN</span>, <span title="SVD">SVD</span>, <span title="MLE">MLE</span>, <span title="QRILC">QRILC</span>, <span title="MinDet">MinDet</span>, <span title="MinProb">MinProb</span>)</code>, default:
                      <code>&#39;Hybrid&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Imputation method to apply:
- 'Hybrid' : Uses an empirical approach (quantile regression) to find a threshold
  below which missing values are imputed according to MNAR and above which values
  are imputed according to MAR. Defaults to mar='KNN' and mnar='QRILC'.
- 'KNN' : Uses K-nearest neighbors to impute missing values under a MAR assumption.
  Defaults to k=15 neighbors.
- 'SVD' : Uses Singular Value Decomposition to impute missing values under a MAR
  assumption. Defaults to k=2 principal components.
- 'MLE' : Uses Maximum Likelihood Estimation (EM algorithm) to impute missing values
  under a MAR assumption.
- 'QRILC' : Uses Quantile Regression for Imputation of Left Censored Data to impute
  missing values under an MNAR assumption. Defaults to tune_sigma=1 (SD of the MNAR
  distribution).
- 'MinDet' : Uses imputation by minimum detected value under an MNAR assumption.
  Defaults to q=0.01 (quantile for minimum value estimation).
- 'MinProb' : Uses imputation by random draws from a Gaussian distribution centered on
  the minimum value. Defaults to q=0.01 and tune_sigma=1.</p>
              </div>
            </li>
            <li>
              <b><code>extra_args</code></b>
                  (<code><span title="dict">dict</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Used in conjunction with methods that take additional parameters.</p>
<p>Valid key-value pairs include:</p>
<ul>
<li>'mar' : {'KNN', 'SVD', 'MLE'} with <code>method='Hybrid'</code>.</li>
<li>'mnar' : {'QRILC', 'MinDet', 'MinProb'} with <code>`method='Hybrid'</code>.</li>
<li>'k' : int
  Number of neighbors (KNN) or principal components (SVD).</li>
<li>'q' : float
  Quantile to estimate minimum value (MinDet, MinProb).</li>
<li>'tune_sigma' : float
  SD of the MNAR distribution (QRILC, MinProb).</li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A <code>SummarizedPy</code> object with imputed missing values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If an invalid value is supplied to 'method'.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Impute missing values using ImputeLCMD's hybrid strategy.
Use example dataset PXD000438 after filtering excessive missingness.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">filter_missingness</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;overall&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;Hybrid&quot;</span><span class="p">)</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.limma_trend_dea" class="doc doc-heading">
            <code class=" language-python"><span class="n">limma_trend_dea</span><span class="p">(</span><span class="n">design_formula</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">contrasts</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">feature_id_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">robust</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">block</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">array_weights</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">extra_args</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.limma_trend_dea" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Run differential expression analysis (DEA) with limma-trend. Limma powers its analyses by
incorporating an empirical mean-variance trend estimated from the data as a prior.
This alleviates the issue of estimating fold changes in the face of heteroscedasticity.
In short, low-abundant features are prone to false positives due to inherently lower variance,
whereas the opposite is true for high-abundant features, which are prone to false negatives.
By modeling the overall mean-variance trend in the data and incorporating it as prior,
information is shared across samples (which powers low-N designs) and features are effectively regularized.
Compared to traditional parametric statistics, this Bayesian approach has consistently been found to be more powerful and
achieve better FDR (false discovery rate) control. Additionally, a robust approximation can be used if the data contain
hypo-/hypervariable features to avoid skewing the mean-variance trend.
By fundamentally utilizing linear models, limma can accommodate complex designs, including fixed and random factors
(i.e. mixed effects, such as nested factors or repeated measures) and their combination
(i.e. to model between- and within-subjects designs). Limma can also incorporate sample quality weights, which are extremely powerful,
especially in noisy datasets, which is often the case with human or animal samples. Using limma's arrayWeights function,
samples are up- or down-weighted based on how variable they are compared to the average sample. Importantly, this function takes
the experimental design into account when estimating the sample weights. Moreover, the user can provide an arbitrary design or none at all
to estimate averaged weights for different groups of samples (i.e. in cases where sample quality is known to be especially poor according to some condition or technical covariate) or
simply estimate sample-specific weights independent of design covariates. arrayWeights is run with the 'REML' method that allows for missing values.
The weights are inversely proportional to sample variability (i.e. a weight of 0.5 is twice as variable as the average sample;
weights &gt;1 are less variable than the average and tend to reflect higher quality). The weights can be stabilized further and squeezed towards 1 by increasing the 'prior_n' parameter &gt;10 (default);
this tends to make weights more symmetric around 1 (average/equality), thus up- and down-weighting samples by similar magnitudes, rather disproportionately up-weighting good samples.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>design_formula</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A formula describing the linear model. Covariates must be present in samples attribute.
Must begin with a tilde (~) and add covariates with '+'.
Note: formula may not contain intercept term</p>
              </div>
            </li>
            <li>
              <b><code>contrasts</code></b>
                  (<code><span title="dict">dict</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A dictionary containing contrast labels (keys) and contrast definitions (values).
Contrasts are defined by adding or subtracting levels of the covariates
included in the design formula. Additional scaling factors are allowed,
such as dividing by the number of included terms to get the average.</p>
              </div>
            </li>
            <li>
              <b><code>feature_id_col</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>The name of a column in the features attribute to name features by.
If None, the method defaults to naming features according to their index.</p>
              </div>
            </li>
            <li>
              <b><code>robust</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to run limma-trend with robust approximation.</p>
              </div>
            </li>
            <li>
              <b><code>block</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Name of a column in samples attribute to using as a blocking variable.
This must be used if running a model with both between- and within-subjects factors.
The blocking variable should correspond to the column (subject)
that gave rise to repeated values.</p>
              </div>
            </li>
            <li>
              <b><code>array_weights</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to estimate sample quality weights.</p>
              </div>
            </li>
            <li>
              <b><code>extra_args</code></b>
                  (<code><span title="dict">dict</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Used in conjunction with array_weights to specify additional arguments.
Valid key-value pairs include:</p>
<ul>
<li>prior_n : int, The number of prior features to add (defaults to 10) to increase squeezing toward 1.</li>
<li>var_group : str, Name of a column in samples indicating groups (levels) that should be assigned different average weights.</li>
<li>sample_id_col : str, A column in samples attribute to use for sample labelling.
  This makes reading the sample weights output in limma_trend_dea.log easier.
  Note that names must be unique and may not start with numbers.
  If None, defaults to naming samples according to their index.</li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A <code>SummarizedPy</code> object with a <code>results</code> attribute containing limma-trend DEA results with columns:</p>
<ul>
<li>contrast_label : name of the contrast</li>
<li>contrast : contrast definition</li>
<li>feature : feature name</li>
<li>logfc : log2 fold change (i.e. regression coefficient)</li>
<li>ci_l : lower confidence interval for logfc</li>
<li>ci_r : upper confidence interval for logfc</li>
<li>aveexpr : average feature expression level</li>
<li>t : t-value for the associated test</li>
<li>p_value : nominal p-value for the associated test</li>
<li>adj_p_val : Benjamini-Hochberg-based false discovery rate</li>
<li>b : log-odds of differential expression</li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If no design_formula or contrasts arguments are provided.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Full DEA pipeline on example dataset PXD000438.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ADC&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">6</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;SCC&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">6</span> <span class="c1"># Add condition variable</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">filter_missingness</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;overall&quot;</span><span class="p">)</span> <span class="c1"># Pre-process</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">transform_features</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;log&quot;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;Hybrid&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">surrogate_variable_analysis</span><span class="p">(</span><span class="n">mod</span><span class="o">=</span><span class="s2">&quot;~condition&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">des</span> <span class="o">=</span> <span class="s2">&quot;~condition+sv_1+sv_2+sv_3&quot;</span> <span class="c1"># design formula (incl. &#39;condition&#39; and surrogate variables)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">contr</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;SCCvsADC&quot;</span><span class="p">:</span> <span class="s2">&quot;SCC-ADC&quot;</span><span class="p">}</span>  <span class="c1"># define contrast (levels must be present in covariates above)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">limma_trend_dea</span><span class="p">(</span><span class="n">design_formula</span><span class="o">=</span><span class="n">des</span><span class="p">,</span> <span class="n">contrasts</span><span class="o">=</span><span class="n">contr</span><span class="p">,</span> <span class="n">array_weights</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># with array_weights option</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span><span class="o">.</span><span class="n">results</span> <span class="c1"># Check newly created results attribute</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.load_example_data" class="doc doc-heading">
            <code class=" language-python"><span class="n">load_example_data</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#depy.summarized_py.SummarizedPy.load_example_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load a real-world example proteomics dataset for demonstration purposes.
The function loads dataset 'PXD000438' from the ImputeLCMD package.
The data were generated from a super-SILAC experiment of human adenocacinoma (ADC) and squamous cell carcinoma (SCC) samples.
The dataset contains six ADC and six SCC samples and 3,709 proteomic features with raw feature intensities and missing values.
Samples 092.1-3 and 441.1-3 are ADC and 561.1-3 and 691.1-3 are SCC.</p>
<p>For more information about the dataset:
https://proteomecentral.proteomexchange.org/cgi/GetDataset?ID=PXD000438</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A <code>SummarizedPy</code> object with 12 samples and 3,709 features.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Load example dataset.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">()</span>
<span class="go">&lt;SummarizedPy(data=ndarray(shape=(3709, 12), dtype=float64), features=DataFrame(shape=(3709, 1)), samples=DataFrame(shape=(12, 1)))&gt;</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.load_sp" class="doc doc-heading">
            <code class=" language-python"><span class="n">load_sp</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#depy.summarized_py.SummarizedPy.load_sp" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load a previously saved <code>SummarizedPy</code> object, stored as a pickle file on disk.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>path</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Path to stored pickle file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A <code>SummarizedPy</code> object.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Load a saved SP object from pickle file.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_sp</span><span class="p">(</span><span class="s2">&quot;my_sp.pkl&quot;</span><span class="p">)</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.plot_pca" class="doc doc-heading">
            <code class=" language-python"><span class="n">plot_pca</span><span class="p">(</span><span class="n">standardize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_comp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fill_by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.plot_pca" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate PCA plot of data using the first two principal components. PCA is computed using scikit-learn's PCA estimator with defaults.
If <code>data</code> contain features with missing values, these will simply be omitted, as PCA requires complete data.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>standardize</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to standardize (i.e. feature-wise z-scoring) data before computing PCA.</p>
              </div>
            </li>
            <li>
              <b><code>n_comp</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Number of principal components to calculate.
Defaults to min(n_samples, n_features) as in sklearn.decimposition.PCA</p>
              </div>
            </li>
            <li>
              <b><code>fill_by</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Valid column name in <code>samples</code> attribute to use for coloring.</p>
              </div>
            </li>
            <li>
              <b><code>label</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to label points according <code>sample</code> variable in <code>samples</code> attribute.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="tuple">tuple</span></code>
              –
              <div class="doc-md-description">
                <p>matplotlib figure and axes objects.</p>
<ul>
<li>fig : <code>.Figure</code></li>
<li>ax : <code>~matplotlib.axes.Axes</code></li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If an invalid value is supplied to any of the arguments.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Generate PCA plot on standardized data.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span><span class="n">standardize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.save_sp" class="doc doc-heading">
            <code class=" language-python"><span class="n">save_sp</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.save_sp" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Save <code>SummarizedPy</code> object to disk using pickle for easy loading in the future.
Method automatically appends '.pkl' to file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>path</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Path to save pickle file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Save SP object to disk.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span><span class="o">.</span><span class="n">save_sp</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;my_sp&quot;</span><span class="p">)</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.select_variable_features" class="doc doc-heading">
            <code class=" language-python"><span class="n">select_variable_features</span><span class="p">(</span><span class="n">top_n</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">top_percentile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.select_variable_features" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Select highly variable features (HVF) based on deviation from data mean-variance trend.
Uses LOWESS to fit a smooth trend to the feature-wise mean and standard deviation values.
Note: if log2 transformation has not been applied using <code>transform_features(method='log',by=2)</code>
it will be applied prior to fitting the mean-variance trend. Data will be returned on the original scale.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>top_n</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Number of top variable features to return.
Mutually exclusive with <code>top_percentile</code>.</p>
              </div>
            </li>
            <li>
              <b><code>top_percentile</code></b>
                  (<code><span title="int">int</span> or <span title="float">float</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>The top Nth percentile (i.e. 100-top_percentile) of variable features to return.
Mutually exclusive with <code>top_n</code>.</p>
              </div>
            </li>
            <li>
              <b><code>plot</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to plot the fitted mean-variance trend and highlight HVFs.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A <code>SummarizedPy</code> object.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If no valid <code>top_n</code> or <code>top_percentile</code> arguments are supplied.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Select top 500 most variable features in example dataset PXD000438 and plot the fitted mean-variance trend.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">select_variable_features</span><span class="p">(</span><span class="n">top_n</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.surrogate_variable_analysis" class="doc doc-heading">
            <code class=" language-python"><span class="n">surrogate_variable_analysis</span><span class="p">(</span><span class="n">mod</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mod0</span><span class="o">=</span><span class="s1">&#39;~1&#39;</span><span class="p">,</span> <span class="n">num_sv</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.surrogate_variable_analysis" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Run surrogate variable analysis to estimate latent factors that capture
expression heterogeneity or hidden batch effects. The surrogate variables (SVs)
will be added to the samples attribute and can be included as covariates in DEA.
SVs are estimated through PCA on the residualized feature matrix after regressing out
known experimental and technical/batch covariates. This is done by supplying the method
with a fully parameterized model (mod), including all known covariates (experimental
and technical; i.e. as present in the samples attribute), and a null model, including
only technical (adjustment) covariates (mod0). The number of significant surrogate variables to estimate can then be specified using the num_sv argument;
alternatively, the method can be run without specifying a number and allowing SVA to estimate the number empirically (using SVA's 'num.sv' function and the default 'leek' method).
Note that this can return 0 SVs and fail. However, it is still possible to find significant SVs by forcing the method to run with a pre-specified num_sv argument.
The mod and mod0 arguments must be specified using R formula formatting, which all start with a tilde (~) symbol and add covariates (+) and their interactions (*). Covariates must be present in the samples attribute.
If no technical covariates are known, the method will run with the recommended default of "~1" (i.e. only using an intercept term).
For more information, see: https://bioconductor.org/packages/3.19/bioc/vignettes/sva/inst/doc/sva.pdf</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mod</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A formula describing the fully parameterized model (incl. all known covariates).
Must begin with a tilde (~) and add covariates with '+'.</p>
              </div>
            </li>
            <li>
              <b><code>mod0</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;~1&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>A formula describing the null model (incl. all known adjustment covariates).
Must begin with a tilde (~) and add covariates with '+'. Defaults to '~1'.</p>
              </div>
            </li>
            <li>
              <b><code>num_sv</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>The number of significant surrogate variables to estimate.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A <code>SummarizedPy</code> object with estimate surrogate variables in the samples attribute.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If no mod formula is supplied.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Use SVA to estimate surrogate variables for inclusion in DEA.
Use example dataset PXD000438: filter missing values and log2 transform features first.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">filter_missingness</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;overall&quot;</span><span class="p">)</span> <span class="c1"># Filter excessive missinginess (this is important)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">transform_features</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;log&quot;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="c1"># Log transform data (important)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;Hybrid&quot;</span><span class="p">)</span> <span class="c1"># Optionally, impute missing remaining values (sva excludes any feature with nan values)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">surrogate_variable_analysis</span><span class="p">(</span><span class="n">mod</span><span class="o">=</span><span class="s2">&quot;~condition&quot;</span><span class="p">)</span> <span class="c1"># Default null model: mod0 = &#39;~1&#39; (intercept-only)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span><span class="o">.</span><span class="n">samples</span>  <span class="c1"># SVs now in samples attribute</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.transform_features" class="doc doc-heading">
            <code class=" language-python"><span class="n">transform_features</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.transform_features" class="headerlink" title="Permanent link">&para;</a></h3>
          <div class="doc-overloads">
<div class="doc-signature codehilite"><pre><span></span><code><span class="nf">transform_features</span><span class="p">(</span>
    <span class="n">method</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">],</span>
    <span class="n">axis</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">by</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span> <span class="s2">&quot;median&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SummarizedPy</span>
</code></pre></div><div class="doc-signature codehilite"><pre><span></span><code><span class="nf">transform_features</span><span class="p">(</span>
    <span class="n">method</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;log&quot;</span><span class="p">],</span> <span class="n">by</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SummarizedPy</span>
</code></pre></div><div class="doc-signature codehilite"><pre><span></span><code><span class="nf">transform_features</span><span class="p">(</span>
    <span class="n">method</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;z-score&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">by</span><span class="p">:</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SummarizedPy</span>
</code></pre></div><div class="doc-signature codehilite"><pre><span></span><code><span class="nf">transform_features</span><span class="p">(</span><span class="n">method</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;vsn&#39;</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">SummarizedPy</span>
</code></pre></div>          </div>


    <div class="doc doc-contents ">

        <p>Mathematically transform features stored in data attribute using one of:
log (base N), center (using mean or median subtraction), or z-score (standardize).</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>axis</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>An axis to perform the transformation along:</p>
<ul>
<li>0 = rows (per-feature)</li>
<li>1 = columns (per-sample).</li>
</ul>
              </div>
            </li>
            <li>
              <b><code>method</code></b>
                  (<code>(<span title="log">log</span>, <span title="center">center</span>, <span title="z">z</span> - <span title="score">score</span>, <span title="vsn">vsn</span>)</code>, default:
                      <code>&#39;log&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Mathematical transformation to apply:</p>
<ul>
<li>'log' : Applies log base N (use <code>by</code> parameter to set base) transformation across entire data array.</li>
<li>'center' : Center data by subtraction (use in conjunction with <code>by</code> parameter).</li>
<li>'z-score' : Standardizes data using z-score transformation (i.e. (x_i-x_mean)/x_std). NB: applies Bessel's N-1 correction to estimate sample standard deviation.</li>
<li>vsn : Applies variance stabilizing normalization, as implemented Huber et al. (2004).</li>
</ul>
              </div>
            </li>
            <li>
              <b><code>by</code></b>
                  (<code><span title="str">str</span> or <span title="int">int</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Used in conjunction with <code>method='center'</code> or <code>method='log'</code>.</p>
<ul>
<li>With 'center' : str, One of 'mean' or 'median' to subtract axis mean or median from each cell.</li>
<li>With 'log' : int, An integer for the base of the logarithm (defaults to 2).</li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><a class="autorefs autorefs-internal" title="SummarizedPy (depy.summarized_py.SummarizedPy)" href="#depy.summarized_py.SummarizedPy">SummarizedPy</a></code>
              –
              <div class="doc-md-description">
                <p>A transformed SummarizedPy object.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If an invalid value is supplied to either axis, method, or by.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Transform feature data in example dataset PXD000438.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">transform_features</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;log&quot;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="c1"># Log transformation (base 2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">transform_features</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;center&quot;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;median&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="c1"># Center data sample-wise by median</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">transform_features</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;z-score&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="c1"># Feature-wise standardization</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">transform_features</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;vsn&quot;</span><span class="p">)</span> <span class="c1"># vsn normalization</span>
</code></pre></div>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="depy.summarized_py.SummarizedPy.volcano_plot" class="doc doc-heading">
            <code class=" language-python"><span class="n">volcano_plot</span><span class="p">(</span><span class="n">contrasts</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">de_colors</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#depy.summarized_py.SummarizedPy.volcano_plot" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate volcano plots for limma-trend results, highlighting the top N up- and downregulated features.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>contrasts</code></b>
                  (<code><span title="list">list</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A list of strings referring to the name of contrasts to plot
(i.e. column 'contrast_label' in results attribute). Defaults to all contrasts.</p>
              </div>
            </li>
            <li>
              <b><code>top_n</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>3</code>
)
              –
              <div class="doc-md-description">
                <p>Number of top up- and down-regulated features to highlight,
ranked by adjusted p-value (FDR).
Note: top_n up- and top_n downregulated features will be displayed,
rather top_n in total. Defaults to top 3.</p>
              </div>
            </li>
            <li>
              <b><code>de_colors</code></b>
                  (<code><span title="dict">dict</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Colors to use for upregulated, downregulated, and non-significant features.
Must supply 'Up', 'Down', and 'ns' as keys with associated colors (str).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="dict">dict</span></code>
              –
              <div class="doc-md-description">
                <p>Dictionaries of matplotlib figure and axes objects for each contrast.</p>
<ul>
<li>fig : <code>.Figure</code></li>
<li>ax : <code>~matplotlib.axes.Axes</code></li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If an invalid value is supplied to either contrasts or de_colors.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Generate volcano plots after running limma_trend_dea method.</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span><span class="o">.</span><span class="n">volcano_plot</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span><span class="o">.</span><span class="n">volcano_plot</span><span class="p">(</span><span class="n">contrasts</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SCCvsADC&quot;</span><span class="p">])</span>
</code></pre></div>


    </div>

</div>



  </div>

    </div>


</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="depy.column_selector"></a>
    <div class="doc doc-contents first">










  <div class="doc doc-children">









<div class="doc doc-object doc-class">



<h2 id="depy.column_selector.ColumnSelector" class="doc doc-heading">
            <code>ColumnSelector</code>


<a href="#depy.column_selector.ColumnSelector" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">



        <p>Object for pre-selecting columns when constructing SummarizedPy from file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>names</code></b>
                  (<code><span title="list">list</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A list of strings matching column names.</p>
              </div>
            </li>
            <li>
              <b><code>regex</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>A string that can be interpreted as a regular expression
by re.search. Note that the search is case-insensitive.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<p><span class="doc-section-title">Examples:</span></p>
    <p>Define columns to import from file. Assume data are in columns labeled "LFQ_intensity_*".</p>
    <div class="codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">depy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">ColumnSelector</span><span class="p">(</span><span class="n">regex</span><span class="o">=</span><span class="s2">&quot;LFQ_intensity_&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">features</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">ColumnSelector</span><span class="p">(</span><span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;proteinID&quot;</span><span class="p">,</span> <span class="s2">&quot;geneSymbol&quot;</span><span class="p">,</span> <span class="s2">&quot;proteinDescription&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sp</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">SummarizedPy</span><span class="p">()</span><span class="o">.</span><span class="n">import_from_delim_file</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;my/path/proteingroups.txt&quot;</span><span class="p">,</span> <span class="n">delim</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">data_selector</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">feature_selector</span><span class="o">=</span><span class="n">features</span><span class="p">)</span>
</code></pre></div>










  <div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="depy.column_selector.ColumnSelector.select_cols" class="doc doc-heading">
            <code class=" language-python"><span class="n">select_cols</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

<a href="#depy.column_selector.ColumnSelector.select_cols" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>df</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>A Pandas <code>DataFrame</code> object on which to perform column selection.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>    DataFrame</code>
              –
              <div class="doc-md-description">
                <p>A pandas <code>DataFrame</code> with selected columns.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>    KeyError</code>
              –
              <div class="doc-md-description">
                <p>If no valid column names are supplied.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>


</div>




  </div>

    </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../usage/" class="btn btn-neutral float-left" title="Usage"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../HISTORY/" class="btn btn-neutral float-right" title="Changelog">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/SebastianDHA/DEPy" class="fa fa-code-fork" style="color: #fcfcfc"> SebastianDHA/DEPy</a>
        </span>
    
    
      <span><a href="../usage/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../HISTORY/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
